src/Context/ContextAwarePluginInterface.php

```diff
@@ -21,10 +21,32 @@
    * already available upon which the definition of subsequent or provided
    * context can be refined.
    *
+   * Note that context gets refined at configuration and execution time of the
+   * plugin.
+   *
    * @param \Drupal\Core\TypedData\DataDefinitionInterface[] $selected_data
    *   An array of data definitions for context that is mapped using a data
    *   selector, keyed by context name.
    */
   public function refineContextDefinitions(array $selected_data);
 
+  /**
+   * Asserts additional metadata for the selected data.
+   *
+   * Allows the plugin to assert additional metadata that is in place when the
+   * plugin has been successfully executed. A typical use-case would be
+   * asserting the node type for a "Node is of type" condition. By doing so,
+   * sub-sequent executed plugins are aware of the metadata and can build upon
+   * it.
+   *
+   * Note that metadata is only asserted on configuration time. The plugin has
+   * to ensure that the run-time data matches the asserted configuration if it
+   * has been executed successfully.
+   *
+   * @param \Drupal\Core\TypedData\DataDefinitionInterface[] $selected_data
+   *   An array of data definitions for context that is mapped using a data
+   *   selector, keyed by context name.
+   */
+  public function assertMetadata(array $selected_data);
```

Comment:
Yeah, that's the one difference - the other is that this is only applied when the plugin is successfully executed. I think "assert metdata" describes it pretty well as the plugin asserts us that there is some additional metadata available if it has been executed.
But yeah, some more docs to help developers to figure out which one to use might be good.. In d7 it was data_info_alter() and metadata_assertions() and they had different parameters, so they locked more different. Maybe we should highlight the difference is in WHAT gets changed:
- refine context definitions: own context definitions
- assert metadata: selected data definitions
--> I'd suggest we make assertMetadata() return changed data definitions, that way we know what changed also. Thinking of https://www.drupal.org/node/2681955 also, we might want require the changed data definitions to be clones also?